Data/Hashable/Class.hs

```diff
@@ -188,14 +206,52 @@ class Hashable a where
     hash = hashWithSalt defaultSalt
 
 #ifdef GENERICS
-    default hashWithSalt :: (Generic a, GHashable (Rep a)) => Int -> a -> Int
-    hashWithSalt salt = ghashWithSalt salt . from
+    default hashWithSalt :: (Generic a, GHashable Zero (Rep a)) => Int -> a -> Int
+    hashWithSalt salt = ghashWithSalt HashArgs0 salt . from
+
+data Zero
+data One
+
+data HashArgs arity a where
+    HashArgs0 :: HashArgs Zero a
+    HashArgs1 :: (Int -> a -> Int) -> HashArgs One a
 
 -- | The class of types that can be generically hashed.
-class GHashable f where
-    ghashWithSalt :: Int -> f a -> Int
+class GHashable arity f where
```

Comment:
Oddly, the changes to GHashable have improved performance. I'll detail this in a top level comment.